ssl/rio/rio_notifier.c

```diff
@@ -0,0 +1,247 @@
+    if (setsockopt(lfd, SOL_SOCKET, SO_EXCLUSIVEADDRUSE, &on, sizeof(on)) < 0) {
+        BIO_closesocket(fd);
+        return -1;
+    }
+
+#else
+# if defined(SOCK_CLOEXEC)
+    socktype |= SOCK_CLOEXEC;
+# endif
+
+    fd = BIO_socket(domain, socktype, protocol, 0);
+    if (fd < 0)
+        return -1;
+
+    /*
+     * Make socket close-on-exec unless this was already done above at socket
+     * creation time.
+     */
+    if (!set_cloexec(fd)) {
+        BIO_closesocket(fd);
+        return -1;
+    }
+#endif
+
+    return fd;
+}
+
+/*
+ * The SOCKET notifier method manually creates a connected TCP socket pair by
+ * temporarily creating a TCP listener on a random port and connecting back to
+ * it.
+ *
+ * Win32 does not support socketpair(2), and Win32 pipes are not compatible with
+ * Winsock select(2). This means our only means of making select(2) wakeable is
+ * to artifically create a loopback TCP connection and send bytes to it.
+ */
+int ossl_rio_notifier_init(RIO_NOTIFIER *nfy)
+{
+    int rc, lfd = -1, rfd = -1, wfd = -1;
+    struct sockaddr_in sa = {0}, accept_sa;
+    socklen_t sa_len = sizeof(sa), accept_sa_len = sizeof(accept_sa);
+
+    /* Create a close-on-exec socket. */
+    lfd = create_socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    if (lfd < 0)
+        return 0;
+
+    /* Bind the socket to a random loopback port. */
+    sa.sin_family       = AF_INET;
+    sa.sin_addr.s_addr  = htonl(INADDR_LOOPBACK);
+    rc = bind(lfd, (const struct sockaddr *)&sa, sizeof(sa));
+    if (rc < 0)
+        goto err;
+
+    /* Determine what random port was allocated. */
+    rc = getsockname(lfd, (struct sockaddr *)&sa, &sa_len);
+    if (rc < 0)
+        goto err;
+
+    /* Start listening. */
+    rc = listen(lfd, 1);
+    if (rc < 0)
+        goto err;
+
+    /* Create another socket to connect to the listener. */
+    wfd = create_socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    if (wfd < 0)
+        goto err;
+
+    /*
+     * Disable Nagle's algorithm on the writer so that wakeups happen
+     * immediately.
+     */
+    if (!BIO_set_tcp_ndelay(wfd, 1))
+        goto err;
+
+    /*
+     * Connect the writer to the listener.
+     */
+    rc = connect(wfd, (struct sockaddr *)&sa, sizeof(sa));
+    if (rc < 0)
+        goto err;
+
+    /*
+     * The connection accepted from the listener is the read side.
+     */
+    rfd = accept(lfd, (struct sockaddr *)&accept_sa, &accept_sa_len);
+    if (rfd < 0)
+        goto err;
+
+    /* Close the listener, which we don't need anymore. */
+    BIO_closesocket(lfd);
+    lfd = -1;
+
+    /*
+     * Sanity check - ensure someone else didn't connect to our listener during
+     * the brief window of possibility above.
+     */
+    if (accept_sa.sin_family != AF_INET || accept_sa.sin_port != sa.sin_port)
+        goto err;
+
+    /* Make both sides of the connection non-blocking. */
+    if (!BIO_socket_nbio(rfd, 1))
```

Comment:
If we're doing non-blocking io on a tcp socket here, wouldn't it be simpler to just use a dgram pair?

Reply:
non-blocking TCP has other semantics than dgram (UDP), yeah?